### ISSUE TITLE
Eliminate TransformState in favour of matrix math

### ISSUE THREAD

**Issue by rdb** [panda3d/panda3d#1690]
## Problem

When users have unoptimised scenes with many objects with unique transformations, the first bottleneck they usually hit is the some element of the transformation cache, usually the garbage collection system. Storing all transforms in a central cache which needs to be managed causes significant overhead.

Even with the cache disabled, the fact that transforms are dynamically allocated and stored on the heap also causes (CPU) cache misses, since every access is a pointer dereference. Allocating and deallocating them causes overhead as well.

Furthermore, the design of a central cache inhibits parallelisation, since a lock needs to be obtained in order to access the cache.

## Analysis

I did some performance tests and found that on modern machines, matrix math is generally just as fast or faster than a cache lookup, depending on the situation. Numbers are in microseconds (lower is better), measured on Linux on an M2 Macbook Air.

| | TransformState | Matrix |
| --- | --- | --- | 
| compose, same xform | 0.0194 | 0.0194 |
| compose, first xform unique | 20.3467 | 0.0123 |
| compose, second xform unique | 15.2843 | 0.0123 |
| compose, same + identity | 0.0071 | 0.0123 |
| compose, unique + identity | 0.0102 | 0.0128 |
| compose, identity + same | 0.0075 | 0.013 |
| compose, identity + unique | 0.009 | 0.0126 |
| compose, both xforms unique | 8.697 | 0.0131 |
| invert, same xform | 0.0266 | 0.0121 |
| invert, unique xform | 105.616 | 0.012 |
| compare, both xforms identical | 0.0004 | 0.0019 |
| compare, both xforms different | 0.0004 | 0.0016 |

The only thing that is measurably slower with matrix math is composing with identity (more on this later), and comparing whether two transforms are the same, which is admittedly the main selling point of having a transform cache, since we can avoid sending a new matrix to the GPU. That said, I feel like at 1 million comparisons, where the 2 ns starts to become a noticeable part of the frame time, you'll have hit far bigger bottlenecks. It's probably worth it making a more realistic stress test to verify the real-world impact.

For composition and inversion, even doing the math over and over again for the same matrix is no slower (within margin of error) than the best-cache scenario for TransformState, hitting the cache every time.

And this is just the basic cache lookups, not taking into account all the additional overhead for managing the cache, such as garbage collection and state removal (which is actually the biggest bottleneck), or concurrent access, all of which which is nonexistent in the matrix math case so cannot be compared.

## Solution

I think this is pretty damning for TransformState and suggests we should get rid of it in favour of directly working with matrices, stored directly on the nodes and not on the heap.

I would suggest that, to keep some level of encapsulation and compatibility, we leave TransformState intact and create a new Transform class which stores a matrix under the hood, but otherwise has a similar interface as TransformState. Nodes would store this new Transform class directly, not as a pointer to the heap. Automatic coercion between a TransformState pointer and Transform is possible, but TransformState would be deprecated in the long run.

Transform could have some additional features over a matrix, such as a flag storing whether it is the identity matrix, so that we can short-circuit that similar to what TransformState does. It hides away the details so that we can change our mind about it later (even switching back to storing a TransformState pointer under the hood if we changed our minds). Caching can still be employed on a case-by-case basis, such as for decomposition, if that turns out to be faster.

We would have to measure the impact of transform comparison, but I suspect that the main area where it is needed (in the renderer), we can have a different way of checking whether two objects have the same transform, using information stored in the scene graph. For instance, when visiting a child node with a transform applied, it may be reasonable to assume that it produces a unique transformation compared to its parent or siblings. If the cull pass assigned a unique number during traversal whenever it encountered a child with a transform, and we compared that number instead of the Transform, we would have something that is likely more than sufficient.

We might have to revisit RenderState in the future as well, but it is less frequently a bottleneck and the RenderState cache advantages are more compelling than that of TransformState.

**Comment by rayanalysis** [comment]
I agree with your technical assessment here that matrices do appear to have a significant performance advantage. I'm not sure there are many downsides to replacing TransformState in such a manner, particularly if the outmoded functionality remains available if needed.

I am in favor of your suggestion to create a new Transform class while keeping TransformState intact.

**Comment by rdb** [comment]
I would also like to consider using column-major 3x4 matrices to store the transforms--this will be mostly an implementation detail hidden behind Transform, except that it excludes projective transforms, which (esoteric exceptions aside) only occur at the projection matrix; it's unheard of for a node to have something other than (0, 0, 0, 1) in the last column.  So this is a 25% reduction of calculations and memory use.

When multiplying with the projection matrix at the end it just gets shuffled back around (we just flip the order of multiplication, actually).

Vulkan / Direct3D ray tracing also requires this format for object-to-world matrices.  On the shaderpipeline branch, we also use it in generated shaders.

**Comment by ArsThaumaturgis** [comment]
Oh wow, this is very interesting to me--especially as I am one of those users who has had trouble with TransformStates!

Provisionally, and based on the information presented here thus far, I'm rather in favour of this change!

(And wow, those are some differences in frame-times in certain cases--100(!) microseconds vs 0.012?!)

I do have a question: You mention keeping TransformState around for compatibility. If this were implemented, and one wanted to switch entirely over to the new system, what would call for changing on the developer-side?

**Comment by rdb** [comment]
Nothing, unless you're dealing with TransformState objects directly, which is unlikely. Otherwise, probably very minor API changes from TransformState to Transform. Details are yet to be worked out, of course.

**Comment by ArsThaumaturgis** [comment]
Okay, in that case I'm very much for it, I believe! ^_^
